File: matlab/postTradeAllocation.m
function optimalAllocation = postTradeAllocation(managedAccounts, symbol1, symbol2, rollDate, filePath, accountName, whichAction, criteriaOption, sanityOnNetPosition, whichOrder)

% af: allocation factor for each account
nAccts = size(managedAccounts,1);
giveUpRef = managedAccounts(:,2);
clearingBroker = managedAccounts(:,3);
af = zeros(1,nAccts);

for j = 1:nAccts
    af(j) = managedAccounts{j,4};
end

switch symbol1
    case {'ty'}
        onePtValue = 1000;
        fileExtension = 'FOUR';
        %ff = 'UST';
    case {'crudeoil'}
        onePtValue = 1000;
        fileExtension = 'FOUR';
        %ff = 'CrudeOil';
    case {'mini'}
        onePtValue = 50;
        fileExtension = 'FOUR';
        %ff = 'Mini';
    case {'bund'}
        onePtValue = 1400;
        fileExtension = 'FOUR';
        %ff = 'Bund';
end

[data, time, buy, sell, price] = filledOrdersManager(symbol1, symbol2, filePath, whichAction, whichOrder);

start = 1;
len1 = size(data,1);
finish = len1;

%disp(['week of ', datestr(x2mdate(time(start))), ' ',datestr(x2mdate(time(finish)))]);

disp(['week of ', datestr(x2mdate(min(time))), ' ',datestr(x2mdate(max(time)))]);

len2 = finish-start+1;

time2 = time(start:finish);
netPosition = zeros(len2,1);
pnl = zeros(len2,1);
cumPnl = zeros(len2,1);
netPosAcct = zeros(len2, nAccts);
cumPnlAcct = zeros(len2, nAccts);
endDayOn = 0;

header = 'Transaction Type,Account,ClOrdID,Sec ID,Side,Qty,TPx,Exec Broker,Clearing Broker,Give-up Ref';
tmpStr = 'D01,052G1135,';
ExecBroker = 'MS';

allocator = cell(nAccts*len2+1,1);
allocator{1} = header;
counter = 1;
fprintf('Length is %d',len1);
for j = 1:len2    
    if j > 1
        lastPeriodPnL = netPosition(j-1)*(price(j)-price(j-1))*onePtValue;
        pnl(j) = lastPeriodPnL;
        cumPnl(j) = cumPnl(j-1) + lastPeriodPnL;
        netPosition(j) = netPosition(j-1) + buy(j) - sell(j);
    else
        lastPeriodPnL = 0;
        pnl(j) = lastPeriodPnL;
        cumPnl(j) = 0;
        netPosition(j) = buy(j) - sell(j);
    end
    
    if j>2
        pnlAcctSoFar = cumPnlAcct(j-1,:);
    else
        pnlAcctSoFar = zeros(1,nAccts);
    end
    %
    if j>1
        if buy(j-1) == 0
            side = -1;
            qty = sell(j-1);
        elseif sell(j-1) == 0
            side = 1;
            qty = buy(j-1);
        end        
        if j>2
            netPosAcct_ = netPosAcct(j-2,:);
        else
            netPosAcct_ = zeros(1, nAccts);
        end
        
        if endDayOn == 1
            optimalAllocation = allocation2;
            endDayOn = 0;
            counter = 1;
            
        else
            [optimalAllocation, parSet] = gridSearch2(cumPnl(j), netPosition(j-1), netPosAcct_, af, qty, pnlAcctSoFar, price(j-1), price(j), onePtValue, side, criteriaOption, sanityOnNetPosition);
            %disp(optimalAllocation*side)
        end
        
        if isempty(rollDate) == 0 
            if floor(datenum(x2mdate(data{j-1,1}))) <  datenum(rollDate)
                symbol = symbol1;
            else
                symbol = symbol2;
            end
        else
            symbol = symbol1;
        end
  
        if side == 1
            for iI = 1:nAccts
                if optimalAllocation(iI) > 0
                    counter = counter+1;
                    allocator{counter} = [tmpStr, data{j-1,2},',' symbol,',BUY,', num2str(optimalAllocation(iI)), ',', num2str(price(j-1), '%.8f'), ',', ExecBroker, ',', clearingBroker{iI,:}, ',', giveUpRef{iI,:}];
                end
            end
        elseif side == -1
            for iI = 1:nAccts
                if optimalAllocation(iI) > 0
                    counter = counter+1;
                    allocator{counter} = [tmpStr, data{j-1,2},',' symbol, ',SELL,', num2str(optimalAllocation(iI)), ',', num2str(price(j-1), '%.8f'), ',', ExecBroker, ',', clearingBroker{iI,:}, ',', giveUpRef{iI,:}];
                end
            end
        end
        
        if j>2
            netPosAcct(j-1,:) = netPosAcct(j-2,:) + side*optimalAllocation;
        else
            netPosAcct(j-1,:) = side*optimalAllocation;
        end
        disp(netPosAcct(j-1,:))
        lastPeriodAcctPnL = netPosAcct(j-1,:)*(price(j)-price(j-1))*onePtValue;
        %disp(['before:' num2str(cumPnlAcct(j,:))]);
        cumPnlAcct(j,:) = cumPnlAcct(j-1,:) + lastPeriodAcctPnL;       
        %disp(['before:' num2str(cumPnlAcct(j,:))]);
    end
    
    if j == len2 %|| floor(x2mdate(time2(j))) ~= floor(x2mdate(time2(j+1)))
        % last trade of the day or last trade of the week
        
        dateString = datestr(x2mdate(time(j)), 'yyyymmdd');
        
        if netPosition(j) == 0
            %(a) if last trade of the day makes the system flat then all 
            % sub-accounts are flat
            scenario = '1';
        elseif netPosition(j-1) == 0
            %(b) if prior to last trade system was flat, allocate according
            %to the allocation factor
            scenario = '2';
        else
            %(c) otherwise we work on optimizing of net position across
            % sub-accounts
            scenario = '3';
        end
        
        if buy(j) == 0
            side = -1;
            qty = sell(j);
        elseif sell(j) == 0
            side = 1;
            qty = buy(j);
        end
        fprintf('\n %d %d',qty, side);
        disp(netPosAcct(j-1,:));
        disp(scenario);
        allocation2 = lastTradeAllocation(qty, side, af, netPosAcct(j-1,:), scenario);
        disp(allocation2)
        if isempty(rollDate) == 0 
            if floor(datenum(x2mdate(data{j,1}))) <  datenum(rollDate)
                symbol = symbol1;
            else
                symbol = symbol2;
            end
        else
            symbol = symbol1;
        end
        
        
        if side == 1
            for iI = 1:nAccts
                if allocation2(iI) > 0
                    counter = counter+1;
                    allocator{counter} = [tmpStr, data{j,2},',' symbol,',BUY,',num2str(allocation2(iI)), ',', num2str(price(j), '%.8f'), ',', ExecBroker, ',', clearingBroker{iI,:}, ',', giveUpRef{iI,:}];
                end
            end
        elseif side == -1
            for iI = 1:nAccts
                if allocation2(iI) > 0
                    counter = counter+1;
                    allocator{counter} = [tmpStr, data{j,2},',' symbol, ',SELL,',num2str(allocation2(iI)), ',', num2str(price(j), '%.8f'), ',', ExecBroker, ',', clearingBroker{iI,:}, ',', giveUpRef{iI,:}];
                end
            end
        end
        endDayOn = 1;
        
        fileToWrite = [lastPeriodPnL cumPnl(len2) netPosition(len2) (netPosAcct(len2-1,:)+side*allocation2) cumPnlAcct(len2,:) price(len2)];
        fileName2 = ['cumPnl_', symbol1,num2str(af(1)), '.csv' ];
        dlmwrite(fileName2, fileToWrite, 'delimiter', ',', 'precision', '%.4f');
        
        time0 = floor(time);
        [time0, m, ~] = unique(time0,'last');
        cumPnlAcct0 = cumPnlAcct(m,:);
        
        output = [0, af; time0, cumPnlAcct0];
        I = isnan(output);
        output(I) = 0;
        I = isinf(output);
        output(I) = 0;
        
        %dlmwrite(['\postTradeAllocation2\assetAllocation\doc\' symbol1, '_pnl_FOUR',num2str(af(1)),'.csv'],output,'precision', '%.6f');
        dlmwrite(['D:\Google Drive\Research&Presentations\postTradeAllocation\' symbol1, '_pnl_FOUR',num2str(af(1)),'.csv'],output,'precision', '%.6f');
        if floor(x2mdate(time2(j))) ~= floor(x2mdate(time2(start)))
            flag_0 = 1;
        else
            flag_0 = 0;
        end
        
        filename = [accountName, '_', fileExtension, '_', dateString, '_', whichAction,'.txt'];
        if exist(filename, 'file') == 0
            fid = fopen(filename, 'wt');
            flag = 0;
        else
            fid = fopen(filename, 'at');
            flag = 1;
        end
        
        if flag == 0
            fprintf(fid, '%s\n', allocator{1});
        end

        if flag_0 == 0
            for i = 2:counter
                fprintf(fid, '%s\n', allocator{i}); 
            end
        elseif flag_0 == 1
            for i = 2:counter
            %for i = 3:counter
                fprintf(fid, '%s\n', allocator{i}); 
            end
        end
        
        fclose(fid);
        
        netPosAcct(len2,:) = netPosAcct(len2-1,:)+side*allocation2;
        
        disp(' ');
        disp('=======================================');
        
        disp('last time and minMae');
        disp([datestr(x2mdate(time2(j))), ' ', num2str(parSet.minMae)]);
        disp('cum p&l (scaled)');
        disp([num2str(cumPnl(j)), ' ', num2str(cumPnlAcct(j,:)./af)]);
        disp('cum p&l (unscaled)');
        disp([num2str(cumPnl(j)), ' ', num2str(cumPnlAcct(j,:))]);
        disp('net position at len2-1');
        disp(num2str(netPosAcct(j-1,:)));
        disp('net position at the end');
        disp(num2str(netPosAcct(len2,:)));
        disp('last two prices');
        disp([num2str(price(j-1)), ' ', num2str(price(j))])
        disp('net position at the fund level');
        disp(num2str(netPosition(j)));
        disp('side multiplied qty');
        disp(num2str(side*qty));
        disp('optimal allocation');
        disp(num2str(allocation2));
        
        disp('-------------------------------------');
        
    end
    %disp(num2str(netPosAcct(j-1,:)));

end

file_id = fopen('position.txt', 'wt');
for j=1:len2
    fprintf(file_id, "%d,%d,%d,%d,%d,%d\n",netPosAcct(j,:));
end
fclose(file_id);

File: postTradeAllocation.cpp
/*
* @Author: shayaan
* @Date:   2020-06-11 16:05:14
* @Last Modified by:   shayaan
* @Last Modified time: 2020-08-24 22:15:31
*/

#include "allInclude.h"
#include "filledOrdersManager.cpp"
#include "parSet.h"
#include "accounts.h"
#include "gridSearch2.cpp"
#include "lastTradeAllocation.cpp"
#include "readOnePointValue.cpp"




static void postTradeAllocation(vector<accountDetails> accounts, string fileName, string symbolFile, vector <double> rollDate,
		string whichAction,char criteriaOption,double sanityOnNetPosition, int whichOrder, string ticker,string output_file_positions, string output_file_returns){

	cout<<"Ticker is: "<<ticker<<endl;
	string symbol;
	int nAccts = accounts.size();
	

	vector<double> af;
	for(int i=0;i<accounts.size();i+=1){
		af.push_back(accounts[i].percentage);
	}

	int onePtValue = 1000; //Default
	onePtValue = readOnePtValue(ticker,symbolFile);
	// onePtValue = 1000; //Default
	cout<<"One Point Value "<<onePtValue<<endl;

	// cout<<"onePtValue: "<<onePtValue<<endl;
	//TODO make changes to this
	
	vector< vector <double> > INPUT = filledOrdersManager(fileName, symbolFile, whichAction, whichOrder);
	vector<double> time = INPUT[0];
	vector<double> buy = INPUT[1];
	vector<double> sell = INPUT[2];
	vector<double> price =  INPUT[3];


	int start = 0;
	int len1 = time.size();
	int finish = start + len1;
	int len2 = finish - start + 1;
	//EXCEL TIME TO UNIX TIME;
	long a = ((long) time[start]-70*365-19)*86400-8*3600;
	long *b = &a;
	long c = ((long) time[finish-1]-70*365-19)*86400-8*3600;
	long *d = &c;
	cout << "From: " << ctime(b) << "To:   " << ctime(d) << endl;
	cout << "Length " << len1<<endl;
	vector <double> netPosition;
	vector <double> pnl ;
	vector <double> cumPnl;
	vector< vector<double> > netPosAcct(nAccts,vector<double>(len1,0));
	vector< vector<double> > cumPnlAcct(nAccts,vector<double>(len1,0));

	vector <double> lastPeriodAcctPnL;




	int endDayOn = 0;

	vector <string> allocator; 
	int counter = 0;
	double lastPeriodPnL;
	vector<double> pnlAcctSoFar;
	int scale;
	int qty;
	vector<double> netPosAcct_(nAccts,0);
	vector <double> optimalAllocation;
	parSet parSet(nAccts,af,onePtValue);

	double temp_sum = 0;
	for(int j = 0; j < len1; j++){

		if(j > 0){

			lastPeriodPnL = netPosition[j-1]*(price[j]-price[j-1])*onePtValue;
			temp_sum += lastPeriodPnL;
			pnl.push_back(lastPeriodPnL);
			cumPnl.push_back(cumPnl[j-1] + lastPeriodPnL);
			netPosition.push_back(netPosition[j-1] + buy[j] - sell[j]);
			// cout<<"J "<<j<<" Net Position:"<<netPosition[j]<<" cumPnl:"<<cumPnl[j]<<" Price:"<<(price[j]-price[j-1])*onePtValue<<endl;
		}

		else{
			lastPeriodPnL = 0;
			pnl.push_back(lastPeriodPnL);
			cumPnl.push_back(0);
			netPosition.push_back(buy[j] - sell[j]);
		}
	   
		if(j > 1){
			pnlAcctSoFar.clear();
			for(int k=0;k<nAccts;k+=1)
				pnlAcctSoFar.push_back(cumPnlAcct[k][j-1]);
		}

		else{
			for(int k=0;k<nAccts;k+=1)
				pnlAcctSoFar.push_back(0);
		}

		if(j>0){

			if(buy[j-1] == 0){
				scale = -1;
				qty = sell[j-1];
			}

			else if(sell[j-1] == 0){
				scale = 1;
				qty = buy[j-1];
			}

			if(j > 1){
				netPosAcct_.clear();
				for(int k=0;k<nAccts;k+=1)
					netPosAcct_.push_back(netPosAcct[k][j-2]);
			}
			
			parSet = gridSearch2(cumPnl[j], netPosition[j-1], netPosAcct_, af, qty, pnlAcctSoFar, price[j-1], price[j], onePtValue, scale, criteriaOption, sanityOnNetPosition);
			optimalAllocation = parSet.optimalAllocation;
			


			if(j > 1){
				for(int k = 0; k < nAccts; k++){
					netPosAcct[k][j-1] = netPosAcct[k][j-2] + scale * optimalAllocation[k];
					// cout<<optimalAllocation[k] * scale<<" ";
				}
			}

			else{
				for(int k = 0; k < nAccts; k++){
					netPosAcct[k][j-1] = scale * optimalAllocation[k]; //All trades have to be in the same direction
					// cout<<optimalAllocation[k] * scale <<" ";
				}

			}

			lastPeriodAcctPnL.clear();
			for(int k = 0; k < nAccts; k++){
				lastPeriodAcctPnL.push_back( netPosAcct[k][j-1]*(price[j]-price[j-1])*onePtValue );
				cumPnlAcct[k][j] = cumPnlAcct[k][j-1] + lastPeriodAcctPnL[k];
			}
		}

		
		if(j>0 and j < len1){
			fstream file;
			file.open(output_file_positions,ios_base::app);
			for(int k=0;k<nAccts;k+=1){
				if(k == nAccts-1)
					file<< netPosAcct[k][j-1];
				else
					file<< netPosAcct[k][j-1]<<",";
				}
			file<<"\n";
			file.close();

			fstream file_returns;
			file_returns.open(output_file_returns,ios_base::app);	
			for(int k=0;k<nAccts;k+=1){
				if(k == nAccts-1)
					file_returns<< cumPnlAcct[k][j-1];
				else
					file_returns<< cumPnlAcct[k][j-1]<<",";
				}
			file_returns<<"\n";
			file_returns.close();
		}
		

		if(j == len1 - 1){

			int scenario;

			if (netPosition[j] == 0){
				scenario = 1;
			}

			else if (netPosition[j-1] == 0){
				scenario = 2;
			}

			else{
				scenario = 3;
			}

			if(buy[j] == 0){
				scale = -1;
				qty = sell[j];
			}else if(sell[j] == 0){
				scale = 1;
				qty = buy[j];
			}

			vector <double> netPosAcctFinal;
			netPosAcctFinal.clear();
			for(int m = 0; m < nAccts; m++){
				netPosAcctFinal.push_back(netPosAcct[m][j - 1]);
			}

			// cout<<"scenario "<<scenario<<endl;

			vector <double> allocation2 = lastTradeAllocation(qty, scale, af, netPosAcctFinal, scenario);
			

			if (rollDate.size() != 0){

			}

			else{
				symbol = fileName;
			}

			cout << "netPosAcct(j,:): ";
			for(int k = 0; k < nAccts; k++){
				netPosAcct[k][j] = netPosAcct[k][j-1] + scale * allocation2[k];
				cout<< netPosAcct[k][j] << " ";
			}
			cout<<endl;

			long e = ((long) time[j]-70*365-19)*86400-8*3600;
			long *f = &e;
			cout << "Time: " << ctime(f) << "minMae: " << parSet.minMae << endl;
			cout << "cumPnl: " << cumPnl[j] << endl;
			cout << "cumPnlAcct: "; 
			
			for(int k=0;k<nAccts;k+=1)
				cout<<cumPnlAcct[k][j]<< " ";
			cout<<endl;

			cout << "PriceLast: " << price[j-1] << endl;
			cout << "PriceNow: " << price[j] << endl;
			cout << "netPosition: " << netPosition[j] << endl;
			 
		}
	
	}

	return;

}
